data/account/participationRegistry.go

```diff
@@ -0,0 +1,947 @@
+			toUpdate = append(toUpdate, record)
+		}
+	}
+	db.mutex.Unlock()
+
+	updated := make(map[ParticipationID]updatingParticipationRecord)
+
+	// Disable active key if there is one
+	for _, record := range toUpdate {
+		record.EffectiveLast = on - 1
+		updated[record.ParticipationID] = updatingParticipationRecord{
+			record.Duplicate(),
+			false,
+		}
+	}
+	// Mark registered.
+	recordToRegister.EffectiveFirst = on
+	recordToRegister.EffectiveLast = recordToRegister.LastValid
+	updated[recordToRegister.ParticipationID] = updatingParticipationRecord{
+		recordToRegister,
+		true,
+	}
+
+	if len(updated) != 0 {
+		db.writeQueue <- partDBWriteRecord{
+			registerUpdated: updated,
+		}
+		db.mutex.Lock()
+		for id, record := range updated {
+			delete(db.dirty, id)
+			db.cache[id] = record.ParticipationRecord
+		}
+		db.mutex.Unlock()
+	}
+
+	db.log.Infof("Registered key (%s) for account (%s) first valid (%d) last valid (%d)\n",
+		id, recordToRegister.Account, recordToRegister.FirstValid, recordToRegister.LastValid)
+	return nil
+}
+
+func (db *participationDB) Record(account basics.Address, round basics.Round, participationAction ParticipationAction) error {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+
+	matches := make([]ParticipationRecord, 0, 1)
+
+	// At most one id should be updated, exit with error if a second is found.
+	for _, record := range db.cache {
+		if record.Account == account && recordActive(record, round) {
+			if len(matches) != 0 {
+				// This probably means there is a bug in the key participation registry Register implementation.
+				return ErrMultipleValidKeys
+			}
+			matches = append(matches, record)
+		}
+	}
+
+	if len(matches) == 0 {
+		// This indicates the participation registry is not synchronized with agreement.
+		return ErrActiveKeyNotFound
+	}
+
+	record := matches[0]
+	// Good case, one key found.
+	switch participationAction {
+	case Vote:
+		record.LastVote = round
+	case BlockProposal:
+		record.LastBlockProposal = round
+	case StateProof:
+		record.LastStateProof = round
+	default:
+		return ErrUnknownParticipationAction
+	}
+
+	db.dirty[record.ParticipationID] = struct{}{}
+	db.cache[record.ParticipationID] = record
+	return nil
+}
+
+// Flush waits until all enqueued asynchronous IO has completed.
+// Waiting for all asynchronous IO to complete includes actions from other threads.
+// Flush waits for the participation registry to be idle.
+// Flush returns the latest error generated by async IO, if any.
+func (db *participationDB) Flush() error {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+
+	db.flushesPending++
+	db.writeQueue <- partDBWriteRecord{
+		flush: true,
+	}
+
+	// See writeThread() for the Broadcast() we're Wait()ing on.
+	db.flushDone.Wait()
+	err := db.asyncErr
+	db.flushesPending--
+	// this allows all waiting Flush()es to see the error that happened
+	if db.flushesPending == 0 {
+		db.asyncErr = nil
+	}
+	return err
+}
+
+func (db *participationDB) Close() {
```

Comment:
We need to add a blocking here until the flushing thread has aborted. Just closing the `writeQueue` isn't enough, as the thread could be still busy doing writing.

Reply:
Fixing with https://github.com/algorand/go-algorand/pull/3191